lib/charms/observability_libs/v0/kubernetes_compute_resources_patch.py

```diff
@@ -468,51 +489,67 @@ def _on_config_changed(self, _):
+        which in that case, will raise the same exception as the original exception that caused it and will put the charm in error state.
+        If the failure is not recoverable (e.g: wrong spec provided), an exception same as the original exception that caused it will be raised and will put the charm in error state.
+        """
         try:
             resource_reqs = self.resource_reqs_func()
             limits = resource_reqs.limits
             requests = resource_reqs.requests
         except ValueError as e:
             msg = f"Failed obtaining resource limit spec: {e}"
-            logger.error(msg)
-            self.on.patch_failed.emit(message=msg)
-            return
+            logger.exception(msg)
+            raise ValueError(msg) from e
 
         for spec in (limits, requests):
             if not is_valid_spec(spec):
                 msg = f"Invalid resource limit spec: {spec}"
                 logger.error(msg)
-                self.on.patch_failed.emit(message=msg)
-                return
+                raise ValueError(msg)
 
         resource_reqs = ResourceRequirements(
             limits=sanitize_resource_spec_dict(limits),  # type: ignore[arg-type]
             requests=sanitize_resource_spec_dict(requests),  # type: ignore[arg-type]
         )
 
         try:
-            self.patcher.apply(resource_reqs)
+            for attempt in tenacity.Retrying(
+                retry=self.PATCH_RETRY_IF,
+                stop=self.PATCH_RETRY_STOP,
+                wait=self.PATCH_RETRY_WAIT,
+                # if you don't succeed raise the last caught exception when you're done
+                reraise=True,
+            ):
+                with attempt:
+                    self._charm.unit.status = MaintenanceStatus(
+                        f"retrying patching resource limit... (attempt #{attempt.retry_state.attempt_number})"
+                    )
+                    self.patcher.apply(resource_reqs)
 
         except exceptions.ConfigError as e:
             msg = f"Error creating k8s client: {e}"
-            logger.error(msg)
-            self.on.patch_failed.emit(message=msg)
-            return
+            logger.exception(msg)
+            raise PatchFailedError(msg) from e
 
         except ApiError as e:
             if e.status.code == 403:
                 msg = f"Kubernetes resources patch failed: `juju trust` this application. {e}"
+
             else:
                 msg = f"Kubernetes resources patch failed: {e}"
 
-            logger.error(msg)
-            self.on.patch_failed.emit(message=msg)
+            logger.exception(msg)
+            raise PatchFailedError(msg) from e
 
         except ValueError as e:
             msg = f"Kubernetes resources patch failed: {e}"
-            logger.error(msg)
-            self.on.patch_failed.emit(message=msg)
+            logger.exception(msg)
+            raise PatchFailedError(msg) from e
```

Comment:
Iiuc, raising the exception here, or in this PR in general, does not convert the status from push to pull.
Seems like the PR added retry which could help, but I don't see how this obviates the need for a stored state.

Reply:
now, you only need to query `is_ready` to get the status of your **successful** patch. The pushing thing was mainly due to failures, right? Now, you don't have to wait for the event of failure, you'll know right away.

Reply:
You mean the charm will crash until the resource patch works?
In that case we don't need `is_ready` at all :D

Reply:
refactored and updated PR description